### ISSUE TITLE
Palette-based PNG output ignores custom ICC profile

### ISSUE THREAD

**Issue by xuanngoc** [lovell/sharp#3392]
### Is this a possible bug in a feature of sharp, unrelated to installation?

- [x] Running `npm install sharp` completes without error.
- [x] Running `node -e "require('sharp')"` completes without error.

If you cannot confirm both of these, please open an [installation issue](https://github.com/lovell/sharp/issues/new?labels=installation&template=installation.md) instead.

### Are you using the latest version of sharp?

- [x] I am using the latest version of `sharp` as reported by `npm view sharp dist-tags.latest`.

If you cannot confirm this, please upgrade to the latest version and try again before opening an issue.

If you are using another package which depends on a version of `sharp` that is not the latest, please open an issue against that package instead.

### What is the output of running `npx envinfo --binaries --system --npmPackages=sharp --npmGlobalPackages=sharp`?

```
  System:
    OS: Linux 5.15 Ubuntu 22.04.1 LTS 22.04.1 LTS (Jammy Jellyfish)
    CPU: (8) x64 11th Gen Intel(R) Core(TM) i5-1135G7 @ 2.40GHz
    Memory: 9.37 GB / 23.27 GB
    Container: Yes
    Shell: 5.8.1 - /usr/bin/zsh
  Binaries:
    Node: 18.2.0 - ~/.nvm/versions/node/v18.2.0/bin/node
    npm: 8.9.0 - ~/.nvm/versions/node/v18.2.0/bin/npm
  npmPackages:
    sharp: ^0.31.1 => 0.31.1 
```
### What are the steps to reproduce?

Create sharp -> extract icc profile from metadata -> create sharp -> resize -> convert to PNG -> withMetadata({icc: iccPath}) -> toBuffer
### What is the expected behaviour?

Output image embedded icc profile 
### Please provide a minimal, standalone code sample, without other dependencies, that demonstrates this problem

```
const path = require("path");
const sharp = require('sharp');
const fs = require('fs');

const file = fs.readFileSync(path.resolve(__dirname, '../../images/metadata/fruit-g236fa3733_1920.png'));

sharp(file).metadata((err, data) => {
  console.log(data)
  const {icc} = data;

  fs.writeFileSync(path.resolve(__dirname, '../../images/metadata/fruit-g236fa3733_1920.icc'), icc)

  sharp(file)
    .resize(300, 300, {fit: 'inside'})
    .png({quality: 100})
    .withMetadata({icc: path.resolve(__dirname, '../../images/metadata/fruit-g236fa3733_1920.icc')})
    .toBuffer((err, data) => {
      console.log(err);
      fs.writeFileSync(path.resolve(__dirname, '../../images/metadata/fruit-g236fa3733_1920_300_icc.png'), data);
    }
  )
})

```
### Please provide sample image(s) that help explain this problem

Original image
![fruit-g236fa3733_1920](https://user-images.githubusercontent.com/32450324/193247504-64b68087-40c3-4daf-bc1c-e99f90d9b074.png)

Output image
![fruit-g236fa3733_1920_300_icc](https://user-images.githubusercontent.com/32450324/193247551-4cb739fc-f84c-4c16-a88d-5decc2052b6c.png)

**Comment by xuanngoc** [comment]
Hello, I would like to report an issue about icc profile. As I remember version 0.30, the image convert to PNG still embeded icc profile. Maybe this issue from new libvips version. Thank you.

**Comment by lovell** [comment]
Is this the same problem as #3390? Are you sure you're using the latest v0.31.1?

**Comment by xuanngoc** [comment]
I sure that I am using the latest v.0.31.1. The previous issue that I created #3390 was fixed. But this is another issue, maybe you missed that in the example code above, I force image output  ```.png({quality: 100})```. The icc miss after execute this operation.

**Comment by xuanngoc** [comment]
I tried use JPG image for input and the output with PNG still miss icc profile. So, I think that this is a new issue.

**Comment by lovell** [comment]
Thanks for clarifying, it is a related but different scenario.

Setting `quality: 100` for PNG will use palette-based output, so the use of ICC profiles doesn't really make sense (although the PNG spec appears to allow it). I guess the previous libpng-based approach used to allow the use of palette+profile together, but the newer libspng doesn't.

https://github.com/randy408/libspng/blob/a4b209fa2cab36b0b7ecd46026dd12c77f66df92/spng/spng.c#L2725-L2727

By not attaching a profile to palette-based PNGs, the latest version should produce more accurate colours, as the quantisation logic assumes sRGB with a gamma of 2.2.

**Comment by kleisauke** [comment]
It looks like a libvips issue, where quantisation is done too early causing metadata and phyiscal pixel dimensions to be disregarded. I've just opened upstream PR https://github.com/libvips/libvips/pull/3073 for this.

**Comment by lovell** [comment]
Ah, I was looking at the decode logic in libspng rather than encode, yes you're quite right Kleis.

I'd probably still try to avoid attaching a profile to a palette-based PNG image. I wonder how well browsers etc. support this?

**Comment by kleisauke** [comment]
The above mentioned code probably checks that the ancillary `iCCP` chunk should appear prior to the critical (and optional) `PLTE` chunk, which corresponds to the [chunk ordering rules in the PNG specification (editor's draft)](https://w3c.github.io/PNG-spec/#5ChunkOrdering).

Testing this palette-based image with the embedded ICC profile seems to work locally in all modern web browsers, as well as Windows photo viewer. `pngcheck` doesn't detects any errors either.

**Comment by lovell** [comment]
sharp v0.31.2 now available with libvips v8.13.3

**Comment by xuanngoc** [comment]
I really appreciate it. I hope I can contribute to sharp or lipvips in the feature. :rocket:
